remote_execution/oss/re_grpc/src/client.rs

```diff
@@ -471,72 +582,183 @@ impl REClient {
+            let metadata = metadata.clone();
+            let fut = async move {
+                // Number of complete (non-partial) messages
+                let num_msgs = size as usize / max_msg_size;
+                let mut segments = vec![];
+                for (i, chunk) in data.chunks(max_msg_size).enumerate() {
+                    segments.push(WriteRequest {
+                        resource_name: resource_name.to_owned(),
+                        write_offset: (i * max_msg_size) as i64,
+                        finish_write: i == num_msgs,
+                        data: chunk.to_owned(),
+                    });
+                }
 
-        let blob_hashes = re_request
-            .requests
-            .iter()
-            .map(|x| x.digest.as_ref().unwrap().hash.clone())
-            .collect::<Vec<String>>();
-        let response = client
-            .batch_update_blobs(with_internal_metadata(re_request, metadata))
-            .await?;
+                let requests = futures::stream::iter(segments);
 
-        let failures: Vec<String> = response
-            .get_ref()
-            .responses
-            .iter()
-            .filter_map(|r| {
-                r.status.as_ref().and_then(|s| {
-                    if s.code == (Code::Ok as i32) {
-                        None
-                    } else {
-                        Some(format!(
-                            "Unable to upload blob '{}', rpc status code: {}, message: \"{}\"",
-                            r.digest.as_ref().map_or("N/A", |d| &d.hash),
-                            s.code,
-                            s.message
-                        ))
+                let resp = bytestream_client
+                    .write(with_internal_metadata(requests, metadata))
+                    .await?;
+                if resp.into_inner().committed_size != size as i64 {
+                    return Err(anyhow::anyhow!(
+                        "Failed to upload inline blob: invalid committed_size from WriteResponse"
+                    ));
+                }
+
+                Ok(vec![hash])
+            };
+            upload_futures.push(Box::pin(fut));
+        }
+
+        // Create futures for any files that needs uploading.
+        for file in request.files_with_digest.unwrap_or_default() {
+            let mut bytestream_client = self.grpc_clients.bytestream_client.clone();
+            let hash = file.digest.hash.clone();
+            let size = file.digest.size_in_bytes;
+            let name = file.name.clone();
+            let metadata = metadata.clone();
+            if size < max_msg_size as i64 {
+                batched_blob_updates.push(BatchUploadRequest::File(file));
+                continue;
+            }
+            let resource_name = format!("uploads/{}/blobs/{}/{}", client_uuid, hash.clone(), size);
+            let fut = async move {
+                let mut file = tokio::fs::File::open(&name).await?;
```

Comment:
It would be good to indicate what we failed to open here in a `with_context` block.